### PR TITLE
logger: improve prefix logger

### LINKED ISSUE #6068: Improve Prefix Logger
## Enhancement

I propose several improvements over the current prefix logger, aims for better readability:

**1. Move the prefix to be the suffix**

We usually use the prefix to provide instance properties, like `<segment_id=10 epoch=5>`, `[region 212]` or even something longer like `MPP<query:436339757161119745,task:1> Table<68>`.

As such prefixes some are short and some are long, it makes us harder to locate the more important thing -- the log message.

We could move these properties to be the suffix, so that it is easy to grep and less disruptive.

Additionally, we could assign a standalone field for them, like:

```diff
- [2022/09/30 10:14:38.766 +08:00] [INFO] [MPPTask.cpp:446] ["MPPTask:MPP<query:436339757161119745,task:1> task ends, time cost is 34 ms."] [thread_id=34]
+ [2022/09/30 10:14:38.766 +08:00] [INFO] [MPPTask.cpp:446] ["task ends, time cost is 34 ms."] [thread_id=34] [info="MPP query:436339757161119745 task:1"]
```

**2. Remove the logger name**

In most cases, our log name is just identical to the source file name:

```
[2022/09/30 10:18:25.285 +08:00] [DEBUG] [PageStorageImpl.cpp:333] ["PageStorage:..."] [thread_id=19]
[2022/09/30 10:18:25.285 +08:00] [INFO] [PageDirectory.cpp:1457] ["PageDirectory:..."] [thread_id=19]
[2022/09/30 10:18:25.285 +08:00] [INFO] [BlobStore.cpp:1044] ["BlobStore:..."] [thread_id=19]
[2022/09/30 10:14:38.743 +08:00] [INFO] [DAGStorageInterpreter.cpp:979] ["DAGStorageInterpreter:..."] [thread_id=34]
[2022/09/30 10:14:38.743 +08:00] [DEBUG] [StorageDeltaMerge.cpp:649] ["StorageDeltaMerge:..."] [thread_id=34]
```

The source file name may be even more informative.

So we could remove most logger names, and only assign log names when it is really needed.

### PR DESCRIPTION
**Pull request by breezewish**
Signed-off-by: Wish <breezewish@outlook.com>

### What problem does this PR solve?

Issue Number: close #6068

### What is changed and how it works?

1. Updated the output layout of the prefix logger. Now it is actually a **suffix logger**, that appends the identifiers at the tail. It also appends the logger name at the tail.

2. Removed most logger names, because it is just duplicated with the file name in the output.

3. Improved performance for `UnifiedLogFormatter` and the `Logger`.

**Examples**:

```c++
auto log = Logger::get();
LOG_INFO(log, "Some messages");
// Output log:
// [INFO] [xxxx.cpp:xxx] ["Some messages"] [thread_id=xx]


auto log = Logger::get("table_id=50", 123);
LOG_INFO(log, "Some messages");
// Output log:
// [INFO] [xxxx.cpp:xxx] ["Some messages"] [source="table_id=50 123"] [thread_id=xx]


auto child = log->getChild("query_id=foo");
LOG_INFO(child, "Some messages from child");
// Output log:
// [INFO] [xxxx.cpp:xxx] ["Some messages from child"] [source="table_id=50 123 query_id=foo"] [thread_id=xx]
```

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test

**Benchmark:**

In my M1 Pro, this PR shows nearly 100% log performance improvement over the master branch, when both using memory log channel:

```text
This PR:
-----------------------------------------------------------------------
Benchmark                             Time             CPU   Iterations
-----------------------------------------------------------------------
UnifiedLogFormatterBM               674 ns          672 ns      1054376
LogBM/PocoLogger_WithoutFmt         563 ns          562 ns      1267634
LogBM/PropsLogger_WithoutFmt        595 ns          595 ns      1182113
LogBM/PocoLogger_FmtShort           675 ns          675 ns      1058169
LogBM/PropsLogger_FmtShort          711 ns          711 ns       984238
LogBM/PocoLogger_FmtLong           1172 ns         1171 ns       596029
LogBM/PropsLogger_FmtLong          1266 ns         1265 ns       554530

Master branch:
-----------------------------------------------------------------------
Benchmark                             Time             CPU   Iterations
-----------------------------------------------------------------------
UnifiedLogFormatterBM              1524 ns         1524 ns       453606
LogBM/PocoLogger_WithoutFmt        1109 ns         1109 ns       634368
LogBM/PropsLogger_WithoutFmt       1270 ns         1270 ns       551259
LogBM/PocoLogger_FmtShort          1278 ns         1278 ns       549869
LogBM/PropsLogger_FmtShort         1395 ns         1394 ns       504632
LogBM/PocoLogger_FmtLong           2058 ns         2057 ns       338282
LogBM/PropsLogger_FmtLong          2115 ns         2115 ns       329790
```

In Intel(R) Xeon(R) CPU E5-2630 v4 @ 2.20GHz server:

```text
This PR:
-----------------------------------------------------------------------
Benchmark                             Time             CPU   Iterations
-----------------------------------------------------------------------
UnifiedLogFormatterBM              1671 ns         1671 ns       415684
LogBM/PocoLogger_WithoutFmt        1405 ns         1405 ns       490317
LogBM/PropsLogger_WithoutFmt       1275 ns         1275 ns       471226
LogBM/PocoLogger_FmtShort          1279 ns         1279 ns       527819
LogBM/PropsLogger_FmtShort         1409 ns         1409 ns       447519
LogBM/PocoLogger_FmtLong           2215 ns         2215 ns       297317
LogBM/PropsLogger_FmtLong          2430 ns         2430 ns       320805

Master branch:
-----------------------------------------------------------------------
Benchmark                             Time             CPU   Iterations
-----------------------------------------------------------------------
UnifiedLogFormatterBM              3210 ns         3210 ns       218866
LogBM/PocoLogger_WithoutFmt        2652 ns         2652 ns       265297
LogBM/PropsLogger_WithoutFmt       2833 ns         2833 ns       244393
LogBM/PocoLogger_FmtShort          2888 ns         2888 ns       245783
LogBM/PropsLogger_FmtShort         3042 ns         3042 ns       225220
LogBM/PocoLogger_FmtLong           4117 ns         4117 ns       167029
LogBM/PropsLogger_FmtLong          4388 ns         4387 ns       157148
```

**Sample log outputs:**

```
[2022/10/05 17:42:25.006 +08:00] [DEBUG] [SSTFilesToBlockInputStream.cpp:180] ["Done loading all kvpairs from [CF=default] [offset=7472] [write_cf_offset=7275] "] [source="table_id=68"] [thread_id=11]
[2022/10/05 17:42:25.018 +08:00] [DEBUG] [SSTFilesToBlockInputStream.cpp:180] ["Done loading all kvpairs from [CF=lock] [offset=197] [write_cf_offset=7275] "] [source="table_id=68"] [thread_id=11]
[2022/10/05 17:42:25.127 +08:00] [DEBUG] [SSTFilesToBlockInputStream.cpp:180] ["Done loading all kvpairs from [CF=default] [offset=7472] [write_cf_offset=7275] "] [source="table_id=68"] [thread_id=11]
[2022/10/05 17:42:25.127 +08:00] [DEBUG] [SSTFilesToBlockInputStream.cpp:180] ["Done loading all kvpairs from [CF=lock] [offset=197] [write_cf_offset=7275] "] [source="table_id=68"] [thread_id=11]
[2022/10/05 17:42:24.943 +08:00] [DEBUG] [PartitionStreams.cpp:506] ["Get schema for table 68"] [source="std::tuple<TableLockHolder, std::shared_ptr<StorageDeltaMerge>, DecodingStorageSchemaSnapshotConstPtr> DB::AtomicGetStorageSchema(const DB::RegionPtr &, DB::TMTContext &)"] [thread_id=9]
[2022/10/05 17:42:24.922 +08:00] [INFO] [ApplySnapshot.cpp:414] ["Begin apply snapshot, new_region=[region 109, applied: term 6 index 7]"] [thread_id=12]
[2022/10/05 17:42:24.922 +08:00] [INFO] [DeltaMergeStore_Ingest.cpp:521] ["Table ingest files - begin, ingest_by_split=false files=[] rows=0 bytes=0 bytes_on_disk=0 range=[0131D088DAC6199996,0131E7AA39958CCCC9) clear=true"] [source="table_id=68"] [thread_id=12]
[2022/10/05 17:42:24.922 +08:00] [INFO] [DeltaMergeStore_Ingest.cpp:585] ["Table ingest files - finished ingested files into segments, ingested_files=[] updated_segments=[<segment_id=38 epoch=0 range=[013139393938643135FF34366339352D3634FF36632D346233642DFF623266652D633839FF3562353834373564FF3400000000000000F9,013330303031306531FF37323032372D3561FF37322D343863392DFF623336612D356537FF3966613133373639FF3800000000000000F8)>] clear=true"] [source="table_id=68"] [thread_id=12]
[2022/10/05 17:42:24.922 +08:00] [DEBUG] [DeltaValueSpace.cpp:159] ["Flush start, delta=MemTableSet: 1 column files, 0 rows, 0 bytes, 1 deletes. ColumnFilePersistedSet [39]: 1 levels, 4 column files, 2 rows, 3334 bytes, 2 deletes."] [source="table_id=68 segment_id=38"] [thread_id=12]
[2022/10/05 17:42:24.923 +08:00] [DEBUG] [ColumnFilePersistedSet.cpp:333] ["ColumnFilePersistedSet [39]: 1 levels, 5 column files, 2 rows, 3334 bytes, 3 deletes., after append 1 column files, level info: [0]: [T_1,D_[01332B7D68EDD99993,013330303031306531FF37323032372D3561FF37322D343863392DFF623336612D356537FF3966613133373639FF3800000000000000F8),T_1,D_[013139393938643135FF34366339352D3634FF36632D346233642DFF623266652D633839FF3562353834373564FF3400000000000000F9,013145C0A1E9666664),D_[0131D088DAC6199996,0131E7AA39958CCCC9)]"] [source="table_id=68 segment_id=38"] [thread_id=12]
[2022/10/05 17:42:28.483 +08:00] [INFO] [KVStore.cpp:325] ["Start to persist [region 97, applied: term 6 index 7], cache size: 0 bytes for `save current region after apply`"] [thread_id=12]
[2022/10/05 17:42:28.483 +08:00] [DEBUG] [KVStore.cpp:327] ["Persist [region 97] done"] [thread_id=12]
[2022/10/05 17:42:28.483 +08:00] [INFO] [ApplySnapshot.cpp:423] ["Finish apply snapshot, new_region=[region 97]"] [thread_id=12]
[2022/10/05 17:42:28.655 +08:00] [DEBUG] [DeltaMergeStore.cpp:134] ["Segment task add to background task pool, segment=<segment_id=53 epoch=2 range=[013930303031306234FF33636362652D3037FF30342D346666342DFF613337662D396131FF3464623333646361FF6300000000000000F8,013939393938356264FF30646233322D3030FF37342D343137332DFF383734652D363836FF6563333161313634FF6600000000000000F9)> task=Flush by_whom=Write"] [source="table_id=68"] [thread_id=20]
[2022/10/05 17:42:28.657 +08:00] [DEBUG] [DeltaMergeStore.cpp:148] ["Segment task pop from background task pool, segment=<segment_id=53 epoch=2 range=[013930303031306234FF33636362652D3037FF30342D346666342DFF613337662D396131FF3464623333646361FF6300000000000000F8,013939393938356264FF30646233322D3030FF37342D343137332DFF383734652D363836FF6563333161313634FF6600000000000000F9)> task=Flush"] [source="table_id=68"] [thread_id=28]
[2022/10/05 17:42:28.657 +08:00] [DEBUG] [DeltaMergeStore_InternalBg.cpp:227] ["Task Flush GC safe point: 0"] [source="table_id=68"] [thread_id=28]
[2022/10/05 17:42:28.657 +08:00] [DEBUG] [DeltaValueSpace.cpp:159] ["Flush start, delta=MemTableSet: 1 column files, 2594 rows, 4319885 bytes, 0 deletes. ColumnFilePersistedSet [54]: 1 levels, 0 column files, 0 rows, 0 bytes, 0 deletes."] [source="table_id=68 segment_id=53"] [thread_id=28]
[2022/10/05 17:42:28.683 +08:00] [DEBUG] [DeltaValueSpace.cpp:192] ["Update index start, delta=<delta_id=54>"] [source="table_id=68 segment_id=53"] [thread_id=28]
[2022/10/05 17:42:28.683 +08:00] [DEBUG] [DeltaValueSpace.cpp:194] ["Update index done, delta=<delta_id=54>"] [source="table_id=68 segment_id=53"] [thread_id=28]
[2022/10/05 17:42:43.579 +08:00] [DEBUG] [MPPTask.cpp:232] ["Handling 100 regions from 1 physical tables in MPP task"] [source=MPP<query:436460051087753217,task:1>] [thread_id=31]
[2022/10/05 17:42:43.653 +08:00] [DEBUG] [MPPTask.cpp:129] ["begin to register the tunnel tunnel1+-1, is_local: false, is_async: true"] [source=MPP<query:436460051087753217,task:1>] [thread_id=31]
[2022/10/05 17:42:43.653 +08:00] [DEBUG] [MPPTask.cpp:294] ["begin to register the task MPP<query:436460051087753217,task:1>"] [source=MPP<query:436460051087753217,task:1>] [thread_id=31]
[2022/10/05 17:42:43.653 +08:00] [INFO] [MPPTaskStatistics.cpp:129] ["{\"query_tso\":436460051087753217,\"task_id\":1,\"is_root\":true,\"sender_executor_id\":\"ExchangeSender_27\",\"executors\":[{\"id\":\"ExchangeSender_27\",\"type\":\"ExchangeSender\",\"children\":[\"HashAgg_8\"],\"outbound_rows\":0,\"outbound_blocks\":0,\"outbound_bytes\":0,\"execution_time_ns\":0,\"partition_num\":1,\"sender_target_task_ids\":[-1],\"exchange_type\":\"PassThrough\",\"connection_details\":[{\"tunnel_id\":\"tunnel1+-1\",\"sender_target_task_id\":-1,\"sender_target_host\":\"127.0.0.1:54373\",\"is_local\":false,\"packets\":0,\"bytes\":0}]},{\"id\":\"HashAgg_8\",\"type\":\"Agg\",\"children\":[\"TableFullScan_25\"],\"outbound_rows\":0,\"outbound_blocks\":0,\"outbound_bytes\":0,\"execution_time_ns\":0},{\"id\":\"TableFullScan_25\",\"type\":\"TableScan\",\"children\":[],\"outbound_rows\":0,\"outbound_blocks\":0,\"outbound_bytes\":0,\"execution_time_ns\":0,\"connection_details\":[{\"is_local\":true,\"packets\":0,\"bytes\":0},{\"is_local\":false,\"packets\":0,\"bytes\":0}]}],\"host\":\"127.0.0.1:3930\",\"task_init_timestamp\":1664962963577953000,\"task_start_timestamp\":0,\"task_end_timestamp\":0,\"compile_start_timestamp\":0,\"compile_end_timestamp\":0,\"read_wait_index_start_timestamp\":0,\"read_wait_index_end_timestamp\":0,\"local_input_bytes\":0,\"remote_input_bytes\":0,\"output_bytes\":0,\"status\":\"INITIALIZING\",\"error_message\":\"\",\"working_time\":0,\"memory_peak\":0}"] [source="mpp_task_tracing MPP<query:436460051087753217,task:1>"] [thread_id=31]
[2022/10/05 17:42:43.653 +08:00] [INFO] [MPPTask.cpp:352] ["task starts preprocessing"] [source=MPP<query:436460051087753217,task:1>] [thread_id=32]
```

**A visual compare of the log change:**

This PR:

image

Master branch:

image

- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by fuzhe1989** [comment]
I'm curious about whether we could treat every log as a static message plus multiple key-value pairs.

**Comment by breezewish** [comment]
> I'm curious about whether we could treat every log as a static message plus multiple key-value pairs.

Definitely we could! It is just not efficient in Poco due to heavy allocations :) For example, adding a single field will +300ns in my M1 (which means roughly 50% slow down)

**Comment by fuzhe1989** [comment]
@breezewish I don't mean it should take effect right now. Maybe first provide a key-value style api is very helpful.

Besides, [spdlog](https://github.com/gabime/spdlog) could be a competitive alternative of Poco log.

**Review by Lloyd-Pottiger — APPROVED** [review]
LGTM

File: dbms/src/Common/FmtUtils.h [hunk 1]
```diff
@@ -14,35 +14,43 @@
...
+
 class FmtBuffer
 {
 public:
     FmtBuffer() = default;
 
-    template <typename... Args>
-    FmtBuffer & fmtAppend(std::string_view fmt, Args &&... args)
+    template <typename FmtStr, typename... Args>
+    inline FmtBuffer & fmtAppend(FmtStr fmt, Args &&... args)
```
Comment (fuzhe1989) [inline dbms/src/Common/FmtUtils.h:29]: methods defined inside of the class body is implicitly inlined.

File: dbms/src/Common/FmtUtils.h [hunk 2]
```diff
@@ -14,35 +14,43 @@
...
     {
-        fmt::vformat_to(std::back_inserter(buffer), fmt, fmt::make_format_args(std::forward<Args>(args)...));
+        fmt::format_to(std::back_inserter(buffer), fmt, std::forward<Args>(args)...);
```
Comment (fuzhe1989) [inline dbms/src/Common/FmtUtils.h:31]: Have you compiled this code under c++20? I'm not sure if `format_to` works well here.

File: dbms/src/Common/FmtUtils.h [hunk 3]
```diff
@@ -14,35 +14,43 @@
...
 
-    FmtBuffer & append(std::string_view s)
+    inline FmtBuffer & append(std::string_view s)
     {
         buffer.append(s.data(), s.data() + s.size());
         return *this;
     }
 
-    std::string toString() const
+    inline FmtBuffer & append(const char & ch)
```
Comment (fuzhe1989) [inline dbms/src/Common/FmtUtils.h:41]: `const char &` -> `char`

File: dbms/src/Common/FmtUtils.h (hunk 1 above)
Comment (breezewish) [reply dbms/src/Common/FmtUtils.h]: Thanks! Lessons learnt 👍

**Review by JaySon-Huang — APPROVED** [review]

**Pull request merged**